### ISSUE TITLE
Issue with/Question around replacing the active asar

### ISSUE THREAD

**Issue by shrewdlogarithm** [atom/electron#2198]
I have a auto-update mechanism for my app - but it relies on being able to replace the active 'asar' archive with a new one.

I'm pretty sure Windows didn't lock the ASAR in early releases - but it does now and this is obviously an issue.

Furthermore - it seems you cannot address the asar itself (or anything named similarly) through the fs package - which I suspect is a bug??

e.g. if fs attempts to access "pathtoapp\resources\app.asar" you get a 'file not found' error (I guess it thinks you're referring to a file within the asar rather than the asar itself?)

e.g. you also cannot access file such as "pathtoapp\resources\app.asar.new" and that seems a clear bug as that's NOT a file within the asar at all???

Basically, we >HAVE< to be able to access the asar file and rename/replace it somehow - and I'm not resorting to command-line scripts or other nonsense.  Apps which can't auto-update are useless in 2015 - apps which need command-line options are not multi-platform - so what's the best way around this then?

**Comment by paulcbetts** [comment]
Most Electron-based apps use [Squirrel.Windows](https://github.com/squirrel/squirrel.windows) to update themselves

**Comment by shrewdlogarithm** [comment]
Seems like a bit of overkill when your actual application is really just one (asar) file and it also seems to be platform specific to Windows which makes it little use to me as EVERYTHING else I've done is completely platform independent (one the the core ideas behind using a Chromium wrappered app surely?)

**Comment by marwanhilmi** [comment]
@shrewdlogarithm The issue you are having with the .asar archive is due the monkey-patched version of the `fs` module - it has been modified in order to support .asar archives however there is a side-effect that other modules which use `fs` will fail if they manipulate .asar archives (ie rimraf on a folder with an asar in it, unzipping zip with an asar in it, etc etc). In order to be able to manipulate .asar files, you need to be sure the modules using `fs` require `original-fs` instead. You can attempt to globally monkey patch `fs` back to original state before using your `fs` modules. This is a known issue marked as `wontfix`.

See:
https://github.com/atom/electron/issues/2035
https://github.com/atom/electron/issues/782
https://github.com/atom/electron/issues/1658

We are implementing a similar strategy to update our application (downloading and replacing an asar), feel free to contact me if you'd like to discuss.

**Comment by shrewdlogarithm** [comment]
AHA! - that makes sense - thanks for that!

**Comment by zcbenz** [comment]
Replacing the `asar` file when app is running will never be supported in Electron, it is like replacing the `.exe` or `.dll` when the program is running, which is very likely to cause troubles.

Only updating the asar file in auto updater sounds an interesting idea, but it can implemented as a node module instead of a built-in feature, and I think it can be a part of the [runtime mode](https://github.com/atom/electron/issues/673) instead of a feature of Electron, so I'm merging this to #673.

**Comment by tomasmcm** [comment]
@shrewdlogarithm I'm also having this issue, can you explain how you were able to overwrite the active app.asar?

**Comment by shrewdlogarithm** [comment]
I never did get a solid working solution - working with scripts to close
the application, swap the file and re-run it would 'sometimes' work.

What's really needed is a system within the executable which looks for an
'upgrade' asar file when it runs and swaps that with the current ASAR - and
a command to restart the current executable really.

Anything else will be - at best - a bodge...

On 22 January 2016 at 17:36, Tomás Correia Marques <notifications@github.com

**Comment by baconface** [comment]
Have you had a look at the [electron-asar-updater](https://github.com/whitesmith/electron-asar-updater) package yet? I have yet to use it myself but it's objective appears to be the same as yours.

**Comment by tomasmcm** [comment]
@baconface  That's because I'm one of the developers of that package.
Apparently the latest version of electron for Windows unlocks the asar file, so we can know overwrite it.

**Comment by baconface** [comment]
Ah okay. Also that package doesn't work based off the example because of a reference to the original package. I put in a PR with the correct reference.

**Comment by valtersboze** [comment]
@tomasmcm did you figure this out? i got as far as downloading the update file. even that required running the app with administrator privileges. but when the asar is downloaded, the app cannot unlink the app.asar (locked) and cannot rename update to anything (permission denied). any workaround? i did manage to get it to replace asar file on linux, but that required running the application from commandline with "sudo" in front. stuck :(
